File: pendulum-vis.cc
#include <memory>

#include "VisualTestViewer.hh"
#include "DoublePendulum.hh"
#include "DoublePendulumEntity.hh"
#include "TwoPendulum.hh"
#include "TwoPendulumEntity.hh"
#include "SinglePendulum.hh"
#include "SinglePendulumEntity.hh"
#include "CameraEntity.hh"

int main(){
  VisualTestViewer viewer(800, 600);

  {
    auto entity = std::make_unique<CameraEntity>();
    entity->SetMode(CameraEntity::PinToSurface);
    entity->SetPosition({10,-10,10});
    entity->SetLookAt({0,40,0});
    entity->SetMovementSpeed(20);
    viewer.add_object(std::move(entity));
  }

  for(int i=0; i<101; i++) {
    DoublePendulum initial;
    initial.length2 = 2;
    initial.mass2 = 1;
    initial.length1 = 4;
    initial.mass1 = 2;
    initial.theta1 = ((100-i)/100.0)*(3.1415926/180);

    auto entity = std::make_unique<DoublePendulumEntity>(initial);
    entity->set_position({-10, i*3, 0});

    viewer.add_object(std::move(entity));
  }

  for(int i=0; i<101; i++) {
    TwoPendulum initial;
    initial.length2 = 2;
    initial.mass2 = 1;
    initial.length1 = 4;
    initial.mass1 = 2;
    initial.theta1 = ((100-i)/100.0)*(3.1415926/180);

    auto entity = std::make_unique<TwoPendulumEntity>(initial);
    entity->set_position({+10, i*3, 0});

    viewer.add_object(std::move(entity));
  }

  for(int i=0; i<101; i++) {
    SinglePendulum initial;
    initial.length = 2;
    initial.theta = ((100-i)/100.0)*(3.1415926/180);

    auto entity = std::make_unique<SinglePendulumEntity>(initial);
    entity->set_position({+30, i*3, 0});

    viewer.add_object(std::move(entity));
  }

  viewer.run();
}

File: pendulum-vis.py
#!/usr/bin/env python3

import math
import os
import sys

sys.path.append(os.path.join(os.path.dirname(__file__), 'lib'))
import Visualizations
import TestCases

#include <memory>

#include "VisualTestViewer.hh"
#include "DoublePendulum.hh"
#include "DoublePendulumEntity.hh"
#include "TwoPendulum.hh"
#include "TwoPendulumEntity.hh"
#include "SinglePendulum.hh"
#include "SinglePendulumEntity.hh"
#include "CameraEntity.hh"

def main():
    viewer = Visualizations.VisualTestViewer(800, 600)

    cam = viewer.make_camera()
    cam.mode = Visualizations.CameraType.PinToSurface
    cam.position = Visualizations.vector3df(10,-10,10)
    cam.look_at(Visualizations.vector3df(0,40,0))
    cam.movement_speed = 20

    for i in range(101):
        initial = TestCases.DoublePendulum()
        initial.length2 = 2
        initial.mass2 = 1
        initial.length1 = 4
        initial.mass1 = 2
        initial.theta1 = math.radians((100-i)/100.0)

        entity = viewer.make_double_pendulum(initial)
        entity.position = Visualizations.vector3df(-10, i*3, 0)

    for i in range(101):
        initial = TestCases.TwoPendulum()
        initial.length2 = 2
        initial.mass2 = 1
        initial.length1 = 4
        initial.mass1 = 2
        initial.theta1 = math.radians((100-i)/100.0)

        entity = viewer.make_two_pendulum(initial)
        entity.position = Visualizations.vector3df(+10, i*3, 0)

    for i in range(101):
        initial = TestCases.SinglePendulum()
        initial.length = 2
        initial.theta = math.radians((100-i)/100.0)

        entity = viewer.make_single_pendulum(initial)
        entity.position = Visualizations.vector3df(+30, i*3, 0);

    viewer.run()


if __name__=='__main__':
    main()
